### PR TITLE
Fix #892: migrated from pytube to  yt-dlp

### LINKED ISSUE #892: Change pytube to yt-dlp for youtube loader
### 🐛 Describe the bug

`pytube` has an issue where it will refuse to download a video saying it is age restricted even though it is not https://github.com/pytube/pytube/issues/1712 . [yt-dlp](https://github.com/yt-dlp/yt-dlp) does not have that problem and actually returns more data about the video, like most replayed segments. Below code has the link to the video that pytube refused to download

```python
from embedchain import Pipeline as App
import os

os.environ["OPENAI_API_KEY"] = "xxx"

app = App()
app.add('https://www.youtube.com/watch?v=4cUtpiDwU0s', data_type='youtube_video')
app.query("What is the video about?")

```
Even this [Langchain tutorial](https://python.langchain.com/docs/integrations/document_loaders/youtube_audio) uses yt-dlp .

It is an easy fix and would require change in this file https://github.com/langchain-ai/langchain/blob/c2d1d903fa35b91018b4d777db2b008fcbaa9fbc/langchain/document_loaders/youtube.py#L59C18-L59C24 . 

Let me know and I can make PR.

### PR DESCRIPTION
**Pull request by MoizKhuzema**
## Description

I addressed issue #892 that states pytube has an issue where it will refuse to download a video saying it is age restricted even though it is not https://github.com/pytube/pytube/issues/1712 . [yt-dlp](https://github.com/yt-dlp/yt-dlp) does not have that problem and actually returns more data about the video. To solve this, i used poetry remove pytube and poetry install yt-dlp to migrate. All tests were passed successfully

Fixes # (issue)
Fix/ #892 

## Type of change
- [x] Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?

used this script for testing: 
from embedchain import Pipeline as App
import os

os.environ["OPENAI_API_KEY"] = "xxx"

app = App()
app.add('https://www.youtube.com/watch?v=4cUtpiDwU0s', data_type='youtube_video')
app.query("What is the video about?")

Please delete options that are not relevant.
- [x] Test Script (please provide)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings

## Maintainer Checklist

- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] Made sure Checks passed

### CODE REVIEW

**Comment by MoizKhuzema** [comment]
if there is an issue with the solution i provided, do let me know and I will make the necessary changes. Thanks

File: pyproject.toml [hunk 1]
```diff
@@ -220,7 +219,5 @@ dropbox = ["dropbox"]
 aws_bedrock = ["boto3"]
 mistralai = ["langchain-mistralai"]
 
-[tool.poetry.group.docs.dependencies]
```
Comment (Dev-Khant) [inline pyproject.toml:223]: @MoizKhuzema No need to remove this for now, we might need it later.

**Comment by Dev-Khant** [comment]
And also make sure to remove pytube from Dataloaders. Then you will have do poetry install again inorder to test it properly.
```
dataloaders=[
    "youtube-transcript-api",
    "docx2txt",
    "duckduckgo-search",
    "pytube",
    "sentence-transformers"
]

**Comment by MoizKhuzema** [comment]
> And also make sure to remove pytube from Dataloaders. Then you will have do poetry install again inorder to test it properly.
> 
> ```
> dataloaders=[
>     "youtube-transcript-api",
>     "docx2txt",
>     "duckduckgo-search",
>     "pytube",
>     "sentence-transformers"
> ]
> ```

@Dev-Khant I removed pytube from dataloaders and re-added [tool.poetry.group.docs.dependencies]. I recompiled the poetry lock file and committed the changes here

**Comment by Dev-Khant** [comment]
@MoizKhuzema  After reinstalling the new poetry changes, are you able to pass all the tests locally?

**Comment by MoizKhuzema** [comment]
> @MoizKhuzema After reinstalling the new poetry changes, are you able to pass all the tests locally?

@Dev-Khant I passed 377 tests, failed 11, and got 4 errors. They are OS errors, Permission errors, and Exception: Model not loaded. None of these tests have anything to do with document loader functionality though which is working properly. I have shared a screenshot that summarizes the tests. If you want me to share more detailed information on the errors/failed tests, do let me know


pic2

**Comment by Dev-Khant** [comment]
Alright sounds good. Will let @deshraj review it once and then it can be merged.
Thanks for contributing @MoizKhuzema!

**Comment by Dev-Khant** [comment]
@MoizKhuzema Can you please resolve the conficts?

**Comment by MoizKhuzema** [comment]
> @MoizKhuzema Can you please resolve the conficts?

@Dev-Khant are you referring to the 11 failed tests?

**Comment by Dev-Khant** [comment]
> > @MoizKhuzema Can you please resolve the conficts?
> 
> @Dev-Khant are you referring to the 11 failed tests?

No `poetry.lock` has conflicts.

**Comment by MoizKhuzema** [comment]
> > > @MoizKhuzema Can you please resolve the conficts?
> > 
> > @Dev-Khant are you referring to the 11 failed tests?
> 
> No `poetry.lock` has conflicts. 

@Dev-Khant I cannot see the conflicts you are referring to. Please guide me a little more

**Comment by Dev-Khant** [comment]
> > > > @MoizKhuzema Can you please resolve the conficts?
> > > 
> > > 
> > > @Dev-Khant are you referring to the 11 failed tests?
> > 
> > 
> > No `poetry.lock` has conflicts.
> 
> @Dev-Khant I cannot see the conflicts you are referring to. Please guide me a little more

Screenshot 2024-06-14 at 11 33 40 AM
Here see this. You can fix this by doing `poetry lock --no-update` on your branch and then pushing your code.

**Comment by MoizKhuzema** [comment]
> > > > > @MoizKhuzema Can you please resolve the conficts?
> > > > 
> > > > 
> > > > @Dev-Khant are you referring to the 11 failed tests?
> > > 
> > > 
> > > No `poetry.lock` has conflicts.
> > 
> > @Dev-Khant I cannot see the conflicts you are referring to. Please guide me a little more
> 
> Screenshot 2024-06-14 at 11 33 40 AM
> Here see this. You can fix this by doing `poetry lock --no-update` on your branch and then pushing your code.

@Dev-Khant I was using the mobile app so I wouldn't see the conflict, my bad.  I can fix this, but later tonight. I'll leave you a message as soon as I push the corrected branch

**Comment by Dev-Khant** [comment]
Sure Thanks @MoizKhuzema, let me know

**Comment by MoizKhuzema** [comment]
> Sure Thanks @MoizKhuzema, let me know

@Dev-Khant I reviewed the conflict. The conflict is because the content-hash of poetry lock file in the branch I committed and the main branch is different. This is because I removed pytube from dataloaders in pyproject.toml and hence from poetry lock when I compile it. The content hash will always be different even after I run  `poetry lock --no-update` on my branch. I checked and there should not be any problems when merging.

**Comment by Dev-Khant** [comment]
> > Sure Thanks @MoizKhuzema, let me know
> 
> @Dev-Khant I reviewed the conflict. The conflict is because the content-hash of poetry lock file in the branch I committed and the main branch is different. This is because I removed pytube from dataloaders in pyproject.toml and hence from poetry lock when I compile it. The content hash will always be different even after I run `poetry lock --no-update` on my branch. I checked and there should not be any problems when merging.
> 
> i found another potential issue in the code. When poetry installs langchain-community, it has a file with path langchain-community/document_loaders/youtube.py and on line 253 it has the following code: `try: from pytube import YouTube
> 
> ```
>     except ImportError:
>         raise ImportError(
>             "Could not import pytube python package. "
>             "Please install it with `pip install pytube`."
>         )`
> ```
> 
> basically langchain-community uses pytube to extract youtube video metadata. If we want to migrate to yt-dlp, we will have to alter that file. Here is the solution proposed by gpt: https://chatgpt.com/share/cbdd5300-5eb4-48a0-a6bc-4f82bc76825c
> 
> can you review the solution proposed and let me know if something like this is plausible. do note this involves making a branch of langchain repo

Thanks for deeply going into the codebase for this issue. Here I don't think it would be a good idea to make change to langchain itself. But what you can do here is see if youtube loader can be changed which would `yt-dlp` only.

**Comment by MoizKhuzema** [comment]
> > > Sure Thanks @MoizKhuzema, let me know
> > 
> > @Dev-Khant I reviewed the conflict. The conflict is because the content-hash of poetry lock file in the branch I committed and the main branch is different. This is because I removed pytube from dataloaders in pyproject.toml and hence from poetry lock when I compile it. The content hash will always be different even after I run `poetry lock --no-update` on my branch. I checked and there should not be any problems when merging.
> > 
> > i found another potential issue in the code. When poetry installs langchain-community, it has a file with path langchain-community/document_loaders/youtube.py and on line 253 it has the following code: `try: from pytube import YouTube
> > 
> > ```
> >     except ImportError:
> >         raise ImportError(
> >             "Could not import pytube python package. "
> >             "Please install it with `pip install pytube`."
> >         )`
> > ```
> > 
> > basically langchain-community uses pytube to extract youtube video metadata. If we want to migrate to yt-dlp, we will have to alter that file. Here is the solution proposed by gpt: https://chatgpt.com/share/cbdd5300-5eb4-48a0-a6bc-4f82bc76825c
> > 
> > can you review the solution proposed and let me know if something like this is plausible. do note this involves making a branch of langchain repo
> 
> Thanks for deeply going into the codebase for this issue. Here I don't think it would be a good idea to make change to langchain itself. But what you can do here is see if youtube loader can be changed which would `yt-dlp` only.

That would require me to change the entire script for the YouTube dataloader itself. I'll make the required changes and propose a working solution

**Comment by Dev-Khant** [comment]
Sure Thanks @MoizKhuzema

**Comment by MoizKhuzema** [comment]
> Sure Thanks @MoizKhuzema

@Dev-Khant I changed the youtube_video.py script to use yt-dlp instead of langchain's youtubeloader that uses pytube. I also changed the test case for youtube loader to reflect this change. I ran all the pytests locally and passed 377 tests, failed 11 tests which were the same ones as before, nothing to do with youtube loader. I did extensive testing to ensure that the code is correct and fault free.

**Comment by Dev-Khant** [comment]
@MoizKhuzema Thanks a lot for working on this and extensively testing your changes. Can you please resolve the conflicts so that we can run tests and review the PR?

**Comment by MoizKhuzema** [comment]
> @MoizKhuzema Thanks a lot for working on this and extensively testing your changes. Can you please resolve the conflicts so that we can run tests and review the PR?

@Dev-Khant conflicts resolved

**Comment by Dev-Khant** [comment]
@MoizKhuzema Can you please fix the linting issue as tests are failing?

**Comment by MoizKhuzema** [comment]
> @MoizKhuzema Can you please fix the linting issue as tests are failing?

@Dev-Khant my bad, i forgot to lint. I'll fix this in a while

**Comment by MoizKhuzema** [comment]
@Dev-Khant I have fixed the lint issue syned the change. Hopefully this time there will be no errors

File: embedchain/loaders/youtube_video.py [hunk 2]
```diff
@@ -7,51 +7,60 @@
...
-    from langchain_community.document_loaders.youtube import _parse_video_id
+    import yt_dlp
 except ImportError:
     raise ImportError(
-        'YouTube video requires extra dependencies. Install with `pip install --upgrade "embedchain[dataloaders]"`'
+        'YouTube video requires extra dependencies. Install with `pip install yt-dlp`'
     ) from None
 from embedchain.helpers.json_serializable import register_deserializable
 from embedchain.loaders.base_loader import BaseLoader
-from embedchain.utils.misc import clean_string
```
Comment (Dev-Khant) [inline embedchain/loaders/youtube_video.py:18]: @MoizKhuzema Why are we not using `clean_string` function anymore?

File: tests/loaders/test_youtube_video.py [hunk 3]
```diff
@@ -1,53 +1,64 @@
...
+        'id': 'VIDEO_ID'
+    }
     mock_transcript = [{"text": "sample text", "start": 0.0, "duration": 5.0}]
+    serialized_transcript = json.dumps(mock_transcript, ensure_ascii=False)  # Serialize the transcript for the test
+    content = ' '.join([item['text'] for item in mock_transcript])
 
-    with patch("embedchain.loaders.youtube_video.YoutubeLoader.from_youtube_url", return_value=mock_loader), patch(
-        "embedchain.loaders.youtube_video.YouTubeTranscriptApi.get_transcript", return_value=mock_transcript
-    ):
+    with patch.object(yt_dlp.YoutubeDL, 'extract_info', return_value=mock_result), \
```
Comment (Dev-Khant) [inline tests/loaders/test_youtube_video.py:34]: Here can you please patch the `YoutubeDL` object from inside `embedchain.loaders.youtube_video. YoutubeVideoLoader.yt_dlp`

File: tests/loaders/test_youtube_video.py [hunk 4]
```diff
@@ -1,53 +1,64 @@
...
+         patch.object(YouTubeTranscriptApi, 'get_transcript', return_value=mock_transcript):
```
Comment (Dev-Khant) [inline tests/loaders/test_youtube_video.py:35]: Same here as well, can you please use `embedchain.loaders.youtube_video. YoutubeVideoLoader. YouTubeTranscriptApi`

File: tests/loaders/test_youtube_video.py [hunk 5]
```diff
@@ -1,53 +1,64 @@
...
+    assert result['doc_id'] == expected_doc_id
+    assert result['data'] == expected_data
 
 def test_load_data_with_empty_doc(youtube_video_loader):
     video_url = "https://www.youtube.com/watch?v=VIDEO_ID"
-    mock_loader = Mock()
-    mock_loader.load.return_value = []
-
-    with patch("embedchain.loaders.youtube_video.YoutubeLoader.from_youtube_url", return_value=mock_loader):
+    with patch.object(yt_dlp.YoutubeDL, 'extract_info', return_value={'entries': []}):
```
Comment (Dev-Khant) [inline tests/loaders/test_youtube_video.py:62]: Same here can you please use `embedchain.loaders.youtube_video. YoutubeVideoLoader.yt_dlp`

**Comment by Dev-Khant** [comment]
@MoizKhuzema Any updates on the comments?

**Comment by MoizKhuzema** [comment]
> @MoizKhuzema Any updates on the comments?

@Dev-Khant I've had engagements these past few days. I'll complete this in a week as soon as I get some time

**Comment by Dev-Khant** [comment]
> > @MoizKhuzema Any updates on the comments?
> 
> @Dev-Khant I've had engagements these past few days. I'll complete this in a week as soon as I get some time

No worries. And also please resolve the conflicts as well. Thanks.

**Comment by taranjeet** [comment]
Hey @MoizKhuzema thanks for your contribution. Closing this issue as there are so many merge conflicts and we have not received any response from you to fix. If you want to resolve, please feel free to open a new PR.

**Pull request closed without merging**